### PR TITLE
Refactoring reorganize taints function in kubectl to expose operations

### LINKED ISSUE #43167: kubectl reporting wrong for taints.
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.3.1185+7e37b895d7c1e9-dirty
Server Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.3.1185+7e37b895d7c1e9-dirty"

**What happened**:
Kubectl is always reporting as node <XYZ> is tainted irrespective of operation when used with local cluster.
./kubectl.sh taint nodes XYZ dedicated-

node "XYZ" **tainted**

**What you expected to happen**:
It should have shown exact operation like tainted, untainted modified and overwritten but because of https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/taint.go#L343. it is always returning tainted.

**How to reproduce it** (as minimally and precisely as possible):
./kubectl.sh taint nodes XYZ dedicated-

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
**What this PR does / why we need it**:
This adds some UX functionality when specifying taints using kubectl.
For example:
```
./kubectl.sh taint nodes XYZ dedicated1=abca2:NoSchedule 
node "XYZ" tainted
 ./kubectl.sh taint nodes XYZ dedicated1=abca1:NoSchedule --overwrite=True
node "XYZ overwritten
./kubectl.sh taint nodes XYZ dedicated1-
node "XYZ" untainted
./kubectl.sh taint nodes XYZ dedicated=abca1:NoSchedule dedicated1-
node "XYZ" modified
```
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #43167 

**Release note**:
```
Fixed the output of kubectl taint node command with minor improvements.
```

### CODE REVIEW

**Comment by ravisantoshgudimetla** [comment]
@smarterclayton PTAL.

**Comment by ethernetdan** [comment]
@k8s-bot ok to test

**Comment by ravisantoshgudimetla** [comment]
@fabianofranz - As I not able to mention sig-cli and sig-cli-pr-reviewers.

**Comment by fabianofranz** [comment]
@kubernetes/sig-cli-pr-reviews

**Comment by ravisantoshgudimetla** [comment]
@fabianofranz I just wanted to make sure if this PR can be reviewed as 1.6 code freeze has officially ended.
P.S:, Should I be a member to mention sig-cli-pr reviewers?

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: can we improve this comment as part of this PR?  the fact that its validating, and reorganizing, ... not sure what 'update' really is doing

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: `s/AddTaints/addTaints` for godoc .  surprised this isnt caught by golint actually...

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: Do you think  we can make this `taintToRemove.Effect != ""` ? Otherwise seems to imply that there can be multiple effects....

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: is this redundant to saying `allErrs` > 0?  if so why not just return `[]error` ?

Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: So, here the errors are different and taints are different, we are returning errors as well as checking if the original length of taints list is equal to old length of taints to check if some of them got deleted.

File: pkg/kubectl/cmd/taint.go
Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: I am just reorganizing in this PR. I will create one more with some other modifications.

File: pkg/kubectl/cmd/taint.go
Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: When can do that but the problem is Effect here is a type again. So this is the convention that others have been using in other locations as well.

Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: can you clarify `Effect here is type again` ?  If `!=`  works it seems like a better comparison.

**Comment by jayunit100** [comment]
lgtm but im not an owner for this section

**Comment by aveshagarwal** [comment]
>This adds some UX functionality when specifying taints using kubectl.
For example:
./kubectl.sh taint nodes XYZ dedicated1=abca2:NoSchedule
node "XYZ" tainted
./kubectl.sh taint nodes XYZ dedicated1=abca1:NoSchedule --overwrite=True
node "XYZ overwritten
./kubectl.sh taint nodes XYZ dedicated1-
node "XYZ" untainted

The above one seem a bit confusing as there might still be other taints left.

>./kubectl.sh taint nodes XYZ dedicated=abca1:NoSchedule dedicated1-
node "XYZ" modified

In general, I am wondering how these UI changes are going to be helpful? May be I am missing something?

Also there are several possibilities I am not sure these UI changes cover all of them:

On one or multiple nodes:
1. Addition of new taints (one or multiple)
2. Deletion of existing taints (one or multiple)
3. Modification/Updates/Overwritten of existing taints (one or multiple)
4. And a combination of above operations.

**Comment by ravisantoshgudimetla** [comment]
So the composite operations(which are complementary like adding taints and deleting them) in a single command will always result in "modified" output, if the composite operations are non-complementary(in this case addition of multiple taints or removal of them) will always result in either "tainted" or "untainted".

If you have atleast one overwritten operation, the output will be "overwritten".

This is something that I came up with but if you have other suggestions, I am more than happy to include.

**Comment by jayunit100** [comment]
@aveshagarwal do you agree that we should provide more feedback ? or you think just saying 'tainted' is enough.  I see your point that some of this is not super informative, but i think its better then just returning 'tainted' every time.

**Comment by davidopp** [comment]
@kubernetes/sig-scheduling-pr-reviews

File: pkg/kubectl/cmd/taint.go
Comment (timothysc) [inline pkg/kubectl/cmd/taint.go]: These should not be here.  
They should either be in a scheduling library or near the api itself.

File: pkg/kubectl/cmd/taint.go
Comment (timothysc) [inline pkg/kubectl/cmd/taint.go]: Again this whole f(n) belongs elsewhere.

Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: You mean somewhere in one of scheduler's packages or api packages? Most of the functions for getting taints are in api/helpers.go. If you want I can make that change but I guess that pre-dates my changes.

Comment (timothysc) [inline pkg/kubectl/cmd/taint.go]: So this should honestly be a topic of conversation for the sig today.  Taints are probably the best example I can think of that has logic sprawl.  

There are changes in api, admission, node, scheduler, controllers, kubectl and TBH it needs to be centralized.  A change in one area will affect another, in it's current implementation I almost consider it a bug-factory. 

/cc @davidopp @gmarek

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: To decide where the right place is, we should decide: what is this describing.

1) If this is describing a kubelet result, then I guess putting them here and making them more descriptive is ok, i.e
```
MODIFICATION="taint was modified"
UNTAINTED="taint was removed"
```
and so on, isolated the semantics, unambiguously to `kubectl`, basically conveying to the user the result of a lower level put/patch/post that may have occured.

**Comment by jayunit100** [comment]
@timothysc 
- agree this could be cleaned up more - but isn't this a step in that direction ? It modularizes the logic while increasing the usability from a user standpoint significantly when adding/removing taints.
- my thoughts would be that it really improves the situation for command line tainting, without increasing technical debt.

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: Agree. We may create a completely separate TaintController controller directory, and move most of the logic there.

Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: Agreed. There are few predicates and priorities in scheduler especially the ones that are coming from nodeSpec or PodSpec that are difficult to develop as they sprawl different components. Till we have a plan to centralize, we need to create a developer document which tells what all files to touch. If not, it becomes difficult for new developers coming in as they have to look at older PR's to find what changes have to be made.

File: pkg/kubectl/cmd/taint.go
Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: @timothysc These are the constants that are being used in output of kubectl, doesn't it make sense to have them here instead of API?
@jayunit100 This gets appended to 'node 'XYZ' from printSuccess at https://github.com/kubernetes/kubernetes/pull/43171/files#diff-92dd749c11309adffd4b2b2112b36303R364. I guess a better way would be to remove 'node 'XYZ' ' part when printing and tell if the taint was added or deleted. I guess that will be a different PR but if you want it, I can include it here.

Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: I think we can all agree these constants shouldnt be exported.

**Comment by adohe** [comment]
@ravisantoshgudimetla @jayunit100 @timothysc @gmarek I would really think this improve command line tainting. And generally we should move `taint` logic to server side, and that should be the direction, like how mesos do [maintenance](https://github.com/apache/mesos/blob/master/docs/maintenance.md).

**Comment by k82cn** [comment]
@timothysc , @jayunit100 , @adohe , @gmarek,

1. agree taint logic could be integrated up more cleanly: moving `taint` logic to server side should be the direction.
2. another suggestion is to have a template for design doc, including feature interaction with cli, UI, the impact to the performance, and so on (we also have a major functions list to review what system looks like with new features). I don-t like "process", but it really helped us to catch up some missed functions before delivered to customer :).

@adohe , should we also update `drain` by `taint`?

**Comment by timothysc** [comment]
So summary would be, we should let this one through and open a new issue on refactor? 

@adohe @k82cn

**Comment by k82cn** [comment]
> So summary would be, we should let this one through and open a new issue on refactor?

Yes, not block this PR :). New PR for refactor and others.

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: Please update the comment here. There's no point of having separate PR for that - we want to keep review load as low as possible.

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: Isn't overwrite a kind of modification? It doesn't seem like it changes logic anywhere, maybe we can just remove it?

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: Wherever we decide to put it, I think we should just use standard verbs: Added, Removed, Modified.

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: It hurts me to see us making copies everywhere... But I don't think this PR can address it.

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: Isn't the latter equivalent to `removed`, i.e. it's always `true` if we reached this part of the code?

**Comment by gmarek** [comment]
Few comments, but on a high level lgtm.

**Comment by adohe** [comment]
```
should we also update drain by taint?
```
@k82cn nop, let's keep this. imho the scope of `taint` is wider than `drain`.

**Comment by adohe** [comment]
and you can see the relationship between `taint` and node drains in [#1574](https://github.com/kubernetes/kubernetes/issues/1574)

File: pkg/kubectl/cmd/taint.go
Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: So, if we are updating one of the existing taint using "--overwrite=True", I am using overwritten flag. If you think, its not worth it. I can take that out.

File: pkg/kubectl/cmd/taint.go
Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: I can take that out but I have to modify the code in calling function, so I did not do it. Probably, as part of rewrite we can clean this up.

File: pkg/kubectl/cmd/taint.go
Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: I guess not, if we do not enter the loop in first place, we cannot return true and 'removed' variable is declared inside loop. So, if we have removed outside loop, then I guess we can.

File: pkg/kubectl/cmd/taint.go
Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: Makes sense...

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: Fair point. I think such change would make the code cleaner.

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: For me this is equivalent to removing "overwritten" one and adding a new in it's place, so I think it's not worth to have this distinction.

File: pkg/kubectl/cmd/taint.go
Comment (ravisantoshgudimetla) [inline pkg/kubectl/cmd/taint.go]: Done.

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: Are we sure we want to return `bool` here?  i do like the clean up with `removed`, but its still unnecessary (normally for kubectl we do this `pkg/kubectl/cmd/util/sanity/cmd_sanity.go:func CheckExamples(cmd *cobra.Command) []error {` as i recall.

**Comment by ravisantoshgudimetla** [comment]
@gmarek PTAL.

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: rather then block this PR ill just let it go :)

**Comment by jayunit100** [comment]
i still dont see any `[]error, bool` returns but no need to block the PR now that its cleaned up.

**Comment by ravisantoshgudimetla** [comment]
> i still dont see any []error, bool returns anywhere in the codebase, but no need to block the PR now that its cleaned up.

@jayunit100, were you referring to any of earlier commits, I guess even now deleteTaints has []error, bool if that is your concern. We need that because this function is being called from reorganizeTaints() and this doesn't need to return any bool values as this called from RunTaints(). The reason that I have an explicit bool return value for deleteTaints() is sometimes, we will not enter the loop and if rely on allErrs alone(by checking on length of errors), we may think that taint has been deleted but in fact it has not been. LMK, if you have other approach.

**Comment by jayunit100** [comment]
@ravisantoshgudimetla its more a personal nit then a technical one: I don't want to block this : I think we press forward with the patch unless others have concerns :), its logically correct as far as i can tell and not introducing anything unmaintainable.

**Comment by jayunit100** [comment]
@timothysc @gmarek PTAL......  looks ok to me and we can iterate on the other things brought up here later.

**Comment by gmarek** [comment]
Oh, sorry - somehow I missed the update.

**Comment by timothysc** [comment]
@ravisantoshgudimetla needs release note.

**Comment by ravisantoshgudimetla** [comment]
@timothysc done.

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: I still don't see fixed comment...

**Comment by ravisantoshgudimetla** [comment]
@gmarek Sorry, did this change on other branch. Now pushed to this one as well. PTAL.

File: pkg/kubectl/cmd/taint.go
Comment (gmarek) [inline pkg/kubectl/cmd/taint.go]: Please be more verbose. "updateTaints adds/removes taints indicated by o (I hate this pattern by the way), while disallowing adding a taint if it would replace existing one if o.overwrite is false.". Or something like that - some native speaker with less than 12 hours on the clock today might have a better suggestion: @timothysc @davidopp

**Review by gmarek — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: @ravisantoshgudimetla `updateTaints applies taints to a node in  a clusters and returns a human understandable description of the resulting operation.`

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: `updateTaints applies a taint operation (o) to a node in a cluster, disallowing certain invalid conditions such as overwriting, and returns a human understandable description summary of the said operation result.`

**Comment by k82cn** [comment]
@ravisantoshgudimetla , can we separate into two commit: 1. your code diff , 2. generated code.

**Comment by ravisantoshgudimetla** [comment]
@k82cn Apologies for the confusion. I was trying to access files from another machine and messed up my git history. There is no generated code in my case as I am touching a single file :).

File: pkg/kubectl/cmd/taint.go
Comment (jayunit100) [inline pkg/kubectl/cmd/taint.go]: `+// updateTaints applies a taint option (o) to a node in the cluster, and after computing the net effect of the operation (i.e. does it result in an overwrite?), it reports the end result back in a way that a user can easily interpret.`

**Comment by jayunit100** [comment]
@gmarek @timothysc @adohe  PTAL , LGTM on my end.

**Comment by jayunit100** [comment]
bump :)

**Review by gmarek — APPROVED** [review]

**Comment by jayunit100** [comment]
Ping @sig-cli-maintainers this one is ready to merge now, reached consensus.

**Pull request merged**
